Data/Functor/Foldable/TH.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE TemplateHaskell #-}
+module Data.Functor.Foldable.TH
+  ( makeBaseFunctor
+  ) where
+
+import Data.Bifunctor (first)
+import Data.Functor.Foldable
+import Language.Haskell.TH
+
+makeBaseFunctor :: Name -> DecsQ
```

Comment:
your 4 points are correct.

Yes, I'll include some kind of `Options` or `Rules` to provide hooks for name generation.

I'm not sure how much different types this should support, e.g. how to generate names for infix constructors (what e.g. `lens` does then?).
I guess supporting `GADTs` isn't good ROI, at least for now.

I'll amend this PR with documentation.